listenbrainz/webserver/static/js/src/PinnedRecordingCard.tsx

```diff
@@ -0,0 +1,240 @@
+  isCurrentUser: Boolean;
+  newAlert: (
+    alertType: AlertType,
+    title: string,
+    message: string | JSX.Element
+  ) => void;
+};
+
+type PinnedRecordingCardState = {
+  currentlyPinned?: Boolean;
+  isDeleted: Boolean;
+};
+
+export default class PinnedRecordingCard extends React.Component<
+  PinnedRecordingCardProps,
+  PinnedRecordingCardState
+> {
+  static contextType = GlobalAppContext;
+  declare context: React.ContextType<typeof GlobalAppContext>;
+
+  constructor(props: PinnedRecordingCardProps) {
+    super(props);
+    const currentlyPinned = this.determineIfCurrentlyPinned();
+    this.state = {
+      currentlyPinned,
+      isDeleted: false,
+    };
+  }
+
+  determineIfCurrentlyPinned = (): Boolean => {
+    const { PinnedRecording } = this.props;
+    const pinnedUntilTime: Date = new Date(PinnedRecording.pinned_until * 1000);
+
+    // invalid date
+    if (Number.isNaN(pinnedUntilTime.getTime())) {
+      return false;
+    }
+    const currentTime: Date = new Date();
+
+    if (currentTime < pinnedUntilTime) {
+      return true;
+    }
+
+    return false;
+  };
+
+  handleError = (error: string | Error, title?: string): void => {
+    const { newAlert } = this.props;
+    if (!error) {
+      return;
+    }
+    newAlert(
+      "danger",
+      title || "Error",
+      typeof error === "object" ? error.message : error
+    );
+  };
+
+  unpinRecording = async () => {
+    const { PinnedRecording, isCurrentUser, newAlert } = this.props;
+    const { APIService, currentUser } = this.context;
+
+    if (isCurrentUser && currentUser?.auth_token) {
+      try {
+        const status = await APIService.unpinRecording(currentUser.auth_token);
+        if (status === 200) {
+          this.setState({ currentlyPinned: false });
+          newAlert(
+            "success",
+            `You unpinned a recording.`,
+            `${PinnedRecording.track_metadata.artist_name} - ${PinnedRecording.track_metadata.track_name}`
+          );
+        }
+      } catch (error) {
+        this.handleError(error, "Error while unpinning recording");
+      }
+    }
+  };
+
+  deletePin = async (pinID: number) => {
+    const { isCurrentUser } = this.props;
+    const { APIService, currentUser } = this.context;
+
+    if (isCurrentUser && currentUser?.auth_token) {
+      try {
+        const status = await APIService.deletePin(
+          currentUser.auth_token,
+          pinID
+        );
+        if (status === 200) {
+          this.setState({ currentlyPinned: false });
+          this.setState({ isDeleted: true });
+        }
+      } catch (error) {
+        this.handleError(error, "Error while deleting pin");
+      }
+    }
+  };
+
+  render() {
+    const { PinnedRecording, isCurrentUser, userName } = this.props;
+    const { currentlyPinned, isDeleted } = this.state;
+    const { track_name } = PinnedRecording.track_metadata;
+    const { artist_name } = PinnedRecording.track_metadata;
+
+    return (
```

Comment:
this component is huge, hundreds of lines of jsx in one component is not ideal. It's not immediately obvious to me if this could be broken into smaller sub-components, but if it could, we should definitely do it. In general small components are easier to reason about and test.

Would you mind taking a look and seeing if this can be broken into smaller subcomponents?

Reply:
The browser and phone components are very similar and only have a few differing lines in render() to shift some html around. I broke up the similarities into JSX-element-returning-functions to make the component more readable. 

Please let me know if this looks better (not sure if you wanted the two split up into different files), if it still looks too big, some other large components are written similarly and I could update those as well.